src/RequestBodyParser.php

```diff
@@ -0,0 +1,125 @@
+    }
+
+    public function withAddedParser(string $mimeType, string $parserClass): self
+    {
+        if (trim($mimeType) === '') {
+            throw new \InvalidArgumentException('The mime type cannot be an empty string.');
+        }
+        if ($parserClass === '') {
+            throw new \InvalidArgumentException('The parser class cannot be an empty string.');
+        }
+        if (strpos($mimeType, '/') === false) {
+            throw new \InvalidArgumentException('Invalid mime type.');
+        }
+        $new = clone $this;
+        $new->parsers[strtolower($mimeType)] = $parserClass;
+        return $new;
+    }
+
+    public function withoutParsers(string ...$mimeTypes): self
+    {
+        $new = clone $this;
+        if (count($mimeTypes) === 0) {
+            $new->parsers = [];
+            return $new;
+        }
+        foreach ($mimeTypes as $mimeType) {
+            if (trim($mimeType) === '') {
+                throw new \InvalidArgumentException('The mime type cannot be an empty string.');
+            }
+            unset($new->parsers[strtolower($mimeType)]);
+        }
+        return $new;
+    }
+
+    public function withBadRequestResponse(): self
+    {
+        $new = clone $this;
+        $new->badRequestResponse = true;
+        return $new;
+    }
+
+    public function withoutBadRequestResponse(): self
+    {
+        $new = clone $this;
+        $new->badRequestResponse = false;
+        return $new;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $contentType = $this->getContentType($request);
+
+        if ($contentType && ($parser = $this->getParser($contentType)) !== null) {
+            try {
+                $parsed = $parser->parse((string)$request->getBody());
+                if (!is_null($parsed) && !is_object($parsed) && !is_array($parsed)) {
+                    throw new \RuntimeException(
+                        'Request body media type parser return value must be an array, an object, or null'
+                    );
+                }
+                $request = $request->withParsedBody($parsed);
+            } catch (ParseException $e) {
+                if ($this->badRequestResponse) {
+                    $response = $this->responseFactory->createResponse(Status::BAD_REQUEST);
+                    $response->getBody()->write(Status::TEXTS[Status::BAD_REQUEST]);
+                    return $response;
+                }
+            }
+        }
+
+        return $handler->handle($request);
+    }
+
+    private function getParser(string $contentType): ?ParserInterface
+    {
+        if (isset($this->parsers[$contentType])) {
+            return $this->container->get($this->parsers[$contentType]);
+        }
+        return null;
+    }
+
+    private function getContentType(ServerRequestInterface $request): ?string
+    {
+        $contentType = $request->getHeaderLine(Header::CONTENT_TYPE);
```

Comment:
Is it possible that a client sends two or more `Content-Type` headers?

Reply:
No.